nixos/modules/services/misc/mjolnir.nix

```diff
@@ -0,0 +1,438 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  cfg = config.services.mjolnir;
+
+  configFile = pkgs.writeText "config.js" ''
+    const fs = require("fs");
+
+    const accessToken = ${if cfg.pantalaimon.enable then
+      "null;"
+      else
+        if cfg.accessTokenFile != null then
+        ''fs.readFileSync("${cfg.accessTokenFile}").toString();''
+        else
+        ''"${cfg.accessToken}";''
+    }
+    const pantalaimonPassword = ${if cfg.pantalaimon.enable then
+      if cfg.pantalaimon.passwordFile != null then
+        ''fs.readFileSync("${cfg.pantalaimon.passwordFile}").toString();''
+        else
+        ''"${cfg.pantalaimon.password}";''
+      else
+        "null;"
+    }
+
+    module.exports = {
+      "homeserverUrl": "${cfg.homeserverUrl}",
+      "dataPath": "${cfg.dataPath}",
+      "autojoinOnlyIfManager": ${boolToString cfg.autojoinOnlyIfManager},
+      "acceptInvitesFromGroup": "${cfg.acceptInvitesFromGroup}",
+      "recordIgnoredInvites": ${boolToString cfg.recordIgnoredInvites},
+      "managementRoom": "${cfg.managementRoom}",
+      "verboseLogging": ${boolToString cfg.verboseLogging},
+      "logLevel": "${cfg.logLevel}",
+      "syncOnStartup": ${boolToString cfg.syncOnStartup},
+      "verifyPermissionsOnStartup": ${boolToString cfg.verifyPermissionsOnStartup},
+      "noop": ${boolToString cfg.noop},
+      "fasterMembershipChecks": ${boolToString cfg.fasterMembershipChecks},
+      "automaticallyRedactForReasons": ${builtins.toJSON cfg.automaticallyRedactForReasons},
+      "protectedRooms": ${builtins.toJSON cfg.protectedRooms},
+      "protectAllJoinedRooms": ${boolToString cfg.protectAllJoinedRooms},
+      "accessToken": accessToken,
+      "pantalaimon": {
+        "use": ${boolToString cfg.pantalaimon.enable},
+        "username": "${cfg.pantalaimon.username}",
+        "password": pantalaimonPassword,
+      },
+      "protections": {
+        "wordlist": {
+          "words": ${builtins.toJSON cfg.protections.wordlist.words},
+        },
+        "minutesBeforeTrusting": ${toString cfg.protections.minutesBeforeTrusting},
+      },
+      "commands": {
+        "allowNoPrefix": ${boolToString cfg.commands.allowNoPrefix},
+        "additionalPrefixes": ${builtins.toJSON cfg.commands.additionalPrefixes},
+        "confirmWildcardBan": ${boolToString cfg.commands.confirmWildcardBan},
+      },
+    };
+  '';
+in
+{
+  options.services.mjolnir = {
+    enable = mkEnableOption "mjolnir";
+
+    homeserverUrl = mkOption {
+      type = types.str;
+      default = "https://matrix.org";
+      description = ''
+        Where the homeserver is located (client-server URL).
+        This should point at pantalaimon if you're using that.
+      '';
+    };
+
+    accessToken = mkOption {
```

Comment:
We don't need this option. If someone is really motivated to store secrets in the nix store they are free to set the value of `accessTokenFile` to something like `pkgs.writeTextFile "badidea.txt" "notsosecret"`.